cachi2/core/package_managers/yarn_classic/resolver.py

```diff
@@ -329,6 +353,36 @@ def resolve_packages(project: Project) -> Iterable[YarnClassicPackage]:
         chain(
             [_get_main_package(project.package_json)],
             _get_workspace_packages(project.source_dir, workspaces),
-            _get_packages_from_lockfile(project.source_dir, yarn_lock, runtime_deps),
+            _get_packages_from_lockfile(project.source_dir, mirror_dir, yarn_lock, runtime_deps),
         )
     )
+
+
+def _read_name_from_tarball(tarball_path: RootedPath) -> str:
+    """Read the package name from the package.json file in the cached tarball."""
+    with tarfile.open(tarball_path) as tar:
+        names = tar.getnames()
+        package_json_subpath = next((n for n in names if n.endswith("package.json")), None)
```

Comment:
A quick experiment tells me that picking the first name that ends with `packages.json` will most likely work. I wonder if this will always be the case? I.e. is it guaranteed that top-level `packages.json` will always come first even if there are nested packages? (This is not a blocker, more of a general question about the tool and I believe it is reasonably safe to just pick the first match)

Reply:
I'm pretty sure that any tarball/git repo that is used as a dependency needs to have its package.json defined in the root folder.